src/Proto3/Wire/Decode.hs

```diff
@@ -145,62 +113,96 @@ data ParsedField = VarintField Word64
-getParsedField Fixed64 =
-    Fixed64Field <$> getByteString 8
-getParsedField LengthDelimited =
-    LengthDelimitedField <$> getLengthDelimited
-
--- | Parse one key/value pair in a protobuf message.
-getKeyVal :: Get (FieldNumber, ParsedField)
-getKeyVal = do
-    (fn, wt) <- getFieldHeader
-    field <- getParsedField wt
-    return (fn, field)
-
--- | Deserializes a protobuf message into a map from field number to all fields
--- labeled with that field number, in their original order.
---
--- This is necessary because of repeated fields, as well as the protobuf
--- requirement that we honor only the last element with a given field number.
---
--- This is as much structure as we can recover without knowing the type of the
--- message.
-getFields :: Get [(FieldNumber, ParsedField)]
-getFields = do
-    keyvals <- many getKeyVal
-    e <- isEmpty
-    unless e $ fail "Encountered bytes that aren't valid key/value pairs."
-    return keyvals
-
 -- | Convert key-value pairs to a map of keys to a sequence of values with that
--- key, in their original occurrence order.
+-- key, in their reverse occurrence order.
 --
--- >>> toMap ([("k1", 3),("k2", 4),("k1", 6)] :: [(String,Int)])
--- fromList [("k1",fromList [3,6]),("k2",fromList [4])]
+-- >>> toMap ([(FieldNumber 1, 3),(FieldNumber 2, 4),(FieldNumber 1, 6)] :: [(FieldNumber,Int)])
+-- fromList [(1,[6,3]),(2,[4])]
 --
--- >>> toMap ([("k2", 7), ("k1", 6), ("k1", 3), ("k2", 4), ("k2", 5)] :: [(String,Int)])
--- fromList [("k1",fromList [6,3]),("k2",fromList [7,4,5])]
---
-toMap :: (Hashable k, Ord k) => [(k, v)] -> M.Map k (Seq v)
-toMap kvs0 = M.fromList (Data.HashMap.Strict.toList hashMap)
-  where
-    kvs1 = map (\(k, v) -> (k, Data.Sequence.singleton v)) kvs0
-    hashMap = Data.HashMap.Strict.fromListWith (flip (<>)) kvs1
+toMap :: [(FieldNumber, v)] -> M.IntMap [v]
+toMap kvs0 = M.fromListWith (<>) . map (fmap (:[])) . map (first (fromIntegral . getFieldNumber)) $ kvs0
+
+decodeWire :: B.ByteString -> Either String [(FieldNumber, ParsedField)]
+decodeWire bstr = drloop bstr []
+ where
+   drloop !bs xs | B.null bs = Right $ reverse xs
+   drloop !bs xs | otherwise = do
+      (w, rest) <- takeVI' bs
+      wt <- gwireType $ fromIntegral (w .&. 7)
+      let fn = w `shiftR` 3
+      (res, rest2) <- takeWT wt rest
+      drloop rest2 ((FieldNumber fn,res):xs)
+
+
+eitherUncons :: B.ByteString -> Either String (Word8, B.ByteString)
+eitherUncons = maybe (Left "failed to parse varint128") Right . B.uncons
+
+
+takeVI' :: B.ByteString -> Either String (Word64, B.ByteString)
+takeVI' !bs =
+  case B.uncons bs of
+     Nothing -> Right (0, B.empty)
+     Just (w1, r1) -> do
+       if w1 < 128 then return (fromIntegral w1, r1) else do
+        let val1 = fromIntegral (w1 .&. 0x7F)
+
+        (w2,r2) <- eitherUncons r1
+        if w2 < 128 then return (val1 .|. (fromIntegral w2 `shiftL` 7), r2) else do
+         let val2 = (val1 .|. (fromIntegral (w2 .&. 0x7F) `shiftL` 7))
```

Comment:
You could replace `.|.` with `+` and `.&. 0x7F` with `- 0x80` to take advantage of x86 `LEA` instructions.  It might also help to put the `fromIntegral` before the `- 0x80`, to avoid compiler-generated `AND` instructions, though I'm not sure if that matters without examining assembly output.

Reply:
ok first two sets of changes made.